src/main/java/org/rumbledb/compiler/VariableDependenciesVisitor.java

```diff
@@ -0,0 +1,527 @@
+import org.rumbledb.expressions.Expression;
+import org.rumbledb.expressions.Node;
+import org.rumbledb.expressions.control.TypeSwitchExpression;
+import org.rumbledb.expressions.control.TypeswitchCase;
+import org.rumbledb.expressions.flowr.CountClause;
+import org.rumbledb.expressions.flowr.ForClause;
+import org.rumbledb.expressions.flowr.GroupByClause;
+import org.rumbledb.expressions.flowr.GroupByVariableDeclaration;
+import org.rumbledb.expressions.flowr.LetClause;
+import org.rumbledb.expressions.flowr.OrderByClause;
+import org.rumbledb.expressions.flowr.OrderByClauseSortingKey;
+import org.rumbledb.expressions.flowr.ReturnClause;
+import org.rumbledb.expressions.flowr.SimpleMapExpression;
+import org.rumbledb.expressions.flowr.WhereClause;
+import org.rumbledb.expressions.module.FunctionDeclaration;
+import org.rumbledb.expressions.module.Prolog;
+import org.rumbledb.expressions.module.VariableDeclaration;
+import org.rumbledb.expressions.postfix.DynamicFunctionCallExpression;
+import org.rumbledb.expressions.postfix.PredicateExpression;
+import org.rumbledb.expressions.primary.ContextItemExpression;
+import org.rumbledb.expressions.primary.FunctionCallExpression;
+import org.rumbledb.expressions.primary.InlineFunctionExpression;
+import org.rumbledb.expressions.primary.NamedFunctionReferenceExpression;
+import org.rumbledb.expressions.primary.VariableReferenceExpression;
+import org.rumbledb.expressions.quantifiers.QuantifiedExpression;
+import org.rumbledb.expressions.quantifiers.QuantifiedExpressionVar;
+
+
+/**
+ * This visitor resolves dependencies between variable and function declarations.
+ * 
+ * If a variable $x depends on a variable $y, then $y must be evaluated before $x.
+ * 
+ * Example:
+ * 
+ * declare variable $x := 1;
+ * declare variable $y := $x;
+ * 
+ * If a variable $x depends on a function f, then f's closure must be built before $x is evaluated.
+ * 
+ * declare function f() { 1 };
+ * declare variable $x := f();
+ * 
+ * If a function f depends on a variable $x, then $x must be evaluated before f's closure is built.
+ *
+ * declare variable $x := 1;
+ * declare function f() { $x };
+ * 
+ * Note that a function cannot depend of a function, as mutually recursive calls are allowed.
+ * 
+ * Once all dependencies have been determined, the visited builds a DAG, builds a topological ordering
+ * thereof, and re-sorts declarations in the prolog for further processing by other visitors.
+ * 
+ */
+public class VariableDependenciesVisitor extends AbstractNodeVisitor<Void> {
+
+    private RumbleRuntimeConfiguration rumbleRuntimeConfiguration;
+
+    /**
+     * Input variable dependencies are lists of variables and functions that an expression depends on.
+     */
+    Map<Node, Set<String>> inputVariableDependencies;
+    /**
+     * Output variable dependencies are lists of variables in the tuples that a clause produces.
+     */
+    Map<Node, Set<String>> outputVariableDependencies;
```

Comment:
I really like the clarity you have added here.

If this variable is indeed used 'only' for clauses, how about we specify this relationship in the variable name instead of a comment?

Reply:
Renamed